### PR TITLE
Add TxDefaultProfile to either EVSE or entire station.

### PR DESCRIPTION
**Pull request by gberardi-pillar**
## Describe your changes

I added some tests for adding a TxDefaultProfile either to the entire station or to a specific EVSE.

Validation is presumed, and eventually validation will be included in add_profile(), but not for this user story.

## Issue ticket number and link

https://github.com/US-JOET/base-camp/issues/98

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [ ] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: lib/ocpp/v201/smart_charging.cpp [hunk 1]
```diff
@@ -192,12 +192,14 @@ SmartChargingHandler::validate_profile_schedules(ChargingProfile& profile,
...
 
-void SmartChargingHandler::add_profile(int32_t evse_id, ChargingProfile& profile) {
+ProfileValidationResultEnum SmartChargingHandler::add_profile(int32_t evse_id, ChargingProfile& profile) {
     if (STATION_WIDE_ID == evse_id) {
         station_wide_charging_profiles.push_back(profile);
     } else {
         charging_profiles[evse_id].push_back(profile);
     }
+
+    return ProfileValidationResultEnum::Valid;
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: I wonder if we should rename this to "Success" or add another variant;

Some other ideas:

* Have `add_profile ()` return a different enum that maps more closely to the OCPP responses and rejection reasons
* Rename `ProfileValidationResultEnum` to something that's also applicable for adding profiles.

File: include/ocpp/v201/smart_charging.hpp [hunk 2]
```diff
@@ -81,7 +81,12 @@ class SmartChargingHandler {
     ///
     /// \brief Adds a given \p profile and associated \p evse_id to our stored list of profiles
     ///
-    void add_profile(int32_t evse_id, ChargingProfile& profile);
+    ProfileValidationResultEnum add_profile(int32_t evse_id, ChargingProfile& profile);
+
+    ///
+    /// \brief Retrieves existing profiles on system.
+    ///
+    ChargingProfile get_profiles();
```
Comment (christopher-davis-afs) [inline include/ocpp/v201/smart_charging.hpp]: This signature seems like it only returns a single profile instead of multiple profiles?

Comment (christopher-davis-afs) [inline include/ocpp/v201/smart_charging.hpp]: Where is the implementation for `get_profiles ()`?

**Review by christopher-davis-afs — CHANGES_REQUESTED** [review]

File: lib/ocpp/v201/smart_charging.cpp (hunk 1 above)
Comment (ghost) [reply lib/ocpp/v201/smart_charging.cpp]: I will second the use of a Success/Reject status for the end result. We can use the return value from validate for any response description/logging.

Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp]: I worry about overloading the one results enum. What about a different one for each step in the assembly line?

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp [hunk 3]
```diff
@@ -545,4 +545,32 @@ TEST_F(ChargepointTestFixtureV201, K01FR49_IfNumberPhasesMissingForACEVSE_ThenSe
...
+TEST_F(ChargepointTestFixtureV201,
+       K01FR14_IfTxDefaultProfileWithSameStackLevelDoesNotExist_ThenApplyStationWideTxDefaultProfileToAllEvses) {
+    create_evse_with_id(DEFAULT_EVSE_ID);
+
+    auto profile = create_charging_profile(DEFAULT_PROFILE_ID, ChargingProfilePurposeEnum::TxDefaultProfile,
+                                           create_charge_schedule(ChargingRateUnitEnum::A), uuid(),
+                                           ChargingProfileKindEnum::Absolute, DEFAULT_STACK_LEVEL);
+
+    auto sut = handler.add_profile(STATION_WIDE_ID, profile);
+    EXPECT_THAT(sut, testing::Eq(ProfileValidationResultEnum::Valid));
```
Comment (folkengine) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp]: Nitpicky spacing thing. My ideal test structure is easily scannable. For me that's a code version of Given, When, Then. All the setup in one section with no empty lines. One section for the when, in this case `auto sut = handler.add_profile(STATION_WIDE_ID, profile);` and then a line feed for all the tests, with no spaces. That way it's easy to scan and instantly isolate on what is actually the subject of the test.

**Review by folkengine — CHANGES_REQUESTED** [review]

File: include/ocpp/v201/smart_charging.hpp (hunk 2 above)
Comment (gberardi-pillar) [reply include/ocpp/v201/smart_charging.hpp]: Ok, so now that I am actually building my code before running tests, I was able to actually test-drive this code AND actually have it exist.

**Review by folkengine — APPROVED** [review]

File: lib/ocpp/v201/smart_charging.cpp [hunk 4]
```diff
@@ -192,12 +192,23 @@ SmartChargingHandler::validate_profile_schedules(ChargingProfile& profile,
...
     } else {
         charging_profiles[evse_id].push_back(profile);
     }
+    SetChargingProfileResponse response;
+    response.status = ChargingProfileStatusEnum::Accepted;
+    return response;
+}
+
+std::vector<ChargingProfile> SmartChargingHandler::get_profiles() {
+    auto all_profiles = station_wide_charging_profiles;
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp:207]: So this is me being unfamiliar with C++, but: is `station_wide_charging_profiles` being copied here, or do modifications to `all_profiles` modify `station_wide_charging_profiles`?

Comment (gberardi-pillar) [reply lib/ocpp/v201/smart_charging.cpp:207]: all_profiles is initialized as a copy of station_wide_charging_profiles. 

The code copies that vector, then adds all of the EVSE-specific profiles.

**Review by christopher-davis-afs — APPROVED** [review]

**Pull request merged**